plugin/pathvirt/pathvirt.cpp

```diff
@@ -0,0 +1,592 @@
+ * is a "physical path".  Internally, DMTCP works with the original "virtual
+ * path" as the canonical name.  But in any system calls, it must translate the
+ * virtual path to the latest "physical path", which will correspond to the
+ * current, post-restart filesystem.
+ */
+static dmtcp::string
+virtual_to_physical_path(const char *virt_path)
+{
+    char *oldPathPtr = NULL;
+    dmtcp::string virtPathString(virt_path?virt_path:"");
+    dmtcp::string physPathString;
+
+    /* quickly return if no swap or NULL path */
+    if (!shouldSwap || !virt_path) {
+        return virtPathString;
+    }
+
+    /* yes, should swap */
+
+    pthread_rwlock_rdlock(&listRwLock);
+    /* check if path is in list of registered paths to swap out */
+    int index = clfind(oldPathPrefixList, virt_path, &oldPathPtr);
+    if (index == -1)
+        return virtPathString;
+
+    /* found it in old list, now get a pointer to the new prefix to swap in*/
+    char *physPathPtr = clget(newPathPrefixList, index);
+    if (physPathPtr == NULL)
+        return virtPathString;
+
+    size_t newElementSz = clgetsize(newPathPrefixList, physPathPtr);
+    size_t oldElementSz = clgetsize(oldPathPrefixList, oldPathPtr);
+
+    /* temporarily null terminate new element */
+    physPathPtr[newElementSz] = '\0';
+
+    /* finally, create full path with the new prefix swapped in */
+    physPathString = physPathPtr;
+    physPathString += "/";
+    physPathString += (virt_path + oldElementSz);
+
+    /* repair the colon list */
+    physPathPtr[newElementSz] = ':';
+
+    pthread_rwlock_unlock(&listRwLock);
+
+    return physPathString;
+}
+
+EXTERNC void
+set_original_path_prefix_list(const char* oldPathPrefix)
+{
+  pthread_rwlock_wrlock(&listRwLock);
+  snprintf(tmpOldPathPrefixList, sizeof(tmpOldPathPrefixList), "%s", oldPathPrefix);
+  tmpBufferModified = true;
+  pthread_rwlock_unlock(&listRwLock);
+}
+
+EXTERNC const char*
+get_original_path_prefix_list()
+{
+  pthread_rwlock_rdlock(&listRwLock);
+  return oldPathPrefixList;
+  pthread_rwlock_unlock(&listRwLock);
+}
+
+EXTERNC const char*
+get_new_path_prefix_list()
+{
+  pthread_rwlock_rdlock(&listRwLock);
+  return newPathPrefixList;
+  pthread_rwlock_unlock(&listRwLock);
+}
+
+/*
+ * DMTCP Setup
+ */
+
+void
+dmtcp_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
+{
+    /* NOTE:  See warning in plugin/README about calls to printf here. */
+    switch (event) {
+    case DMTCP_EVENT_INIT:
+    {
+        /* On init, check if they've specified paths to virtualize via
+           DMTCP_PATH_PREFIX env */
+        char *oldEnv = getenv(ENV_ORIG_DPP);
+        if (oldEnv) {
+            /* if so, save it to buffer */
+            snprintf(oldPathPrefixList, sizeof(oldPathPrefixList), "%s", oldEnv);
+        }
+        pthread_rwlock_init(&listRwLock, NULL);
+        break;
+    }
+    case DMTCP_EVENT_RESTART:
+    {
+        JTRACE("\n*** The plugin %s is being called after restart. ***");
```

Comment:
I suppose we can delete these now, they were just from a plugin I copied when initially writing this.